main/chessy.c

```diff
@@ -1,136 +1,294 @@
+{
+    ESP_LOGI(TAG, "Initializing hall effect sensors");
+    gpio_config_t io_conf = {
+        .intr_type = GPIO_INTR_DISABLE,
+        .mode = GPIO_MODE_OUTPUT,
+        .pin_bit_mask = (1ULL << HALL_COL_SW1) | (1ULL << HALL_COL_SW2) | 
+                       (1ULL << HALL_COL_SW3) | (1ULL << HALL_COL_SW4) |
+                       (1ULL << HALL_COL_SW5) | (1ULL << HALL_COL_SW6) |
+                       (1ULL << HALL_COL_SW7) | (1ULL << HALL_COL_SW8),
+    };
+    ESP_ERROR_CHECK(gpio_config(&io_conf));
+
+    io_conf.mode = GPIO_MODE_INPUT;
+    io_conf.pin_bit_mask = (1ULL << HALL_ROW1) | (1ULL << HALL_ROW2) |
+                          (1ULL << HALL_ROW3) | (1ULL << HALL_ROW4) |
+                          (1ULL << HALL_ROW5) | (1ULL << HALL_ROW6) |
+                          (1ULL << HALL_ROW7) | (1ULL << HALL_ROW8);
+    ESP_ERROR_CHECK(gpio_config(&io_conf));
+}
+
+// Read hall effect sensors
+void hall_read(uint8_t matrix[ROW_NUM][COL_NUM])
+{
+    static uint8_t col_pins[] = {
+        HALL_COL_SW1, HALL_COL_SW2, HALL_COL_SW3, HALL_COL_SW4,
+        HALL_COL_SW5, HALL_COL_SW6, HALL_COL_SW7, HALL_COL_SW8
+    };
+    
+    static uint8_t row_pins[] = {
+        HALL_ROW1, HALL_ROW2, HALL_ROW3, HALL_ROW4,
+        HALL_ROW5, HALL_ROW6, HALL_ROW7, HALL_ROW8
+    };
+
+    // Clear matrix first
+    memset(matrix, 0, ROW_NUM * COL_NUM);
+
+    // Scan each column
+    for (int col = 0; col < COL_NUM; col++) {
+        gpio_set_level(col_pins[col], 1);
+        vTaskDelay(pdMS_TO_TICKS(10));
+        
+        // Read each row
+        for (int row = 0; row < ROW_NUM; row++) {
+            matrix[row][col] = gpio_get_level(row_pins[row]);
+        }
+        
+        gpio_set_level(col_pins[col], 0);
     }
+}
 
-    if (strcmp(line, "p") == 0) {
-        print_board();
-        linenoiseFree(line);
-        return false;
-    } else if (strcmp(line, "m") == 0) {
-        print_move_list();
-        linenoiseFree(line);
-        return false;
-    } else if (strcmp(line, "r") == 0) {
-        init_board();
-        print_board();
-        linenoiseFree(line);
-        return false;
-    } else if (strcmp(line, "q") == 0) {
-        linenoiseFree(line);
-        exit(0);
+// Verify that the physical board matches the expected state
+static bool verify_board_state(const char board[8][8], const uint8_t matrix[ROW_NUM][COL_NUM])
+{
+    for (int row = 0; row < ROW_NUM; row++) {
+        for (int col = 0; col < COL_NUM; col++) {
+            // If there's a piece in the board array, there should be a magnet detected
+            if (board[row][col] != ' ' && !matrix[row][col]) {
+                printf("Error: Missing piece at %c%d\n", 'a' + col, 8 - row);
+                led_set(col, row, COLOR_ERROR);
+                led_refresh();
+                return false;
+            }
+            // If there's no piece in the board array, there should be no magnet detected
+            if (board[row][col] == ' ' && matrix[row][col]) {
```

Comment:
The value from the hall sensor is 1 if there is nothing detected, 0 if there is something. So the logic here should be reversed. Or we should change this when reading the data from the hall sensors.
```suggestion
            if (board[row][col] != ' ' && matrix[row][col]) {
                printf("Error: Missing piece at %c%d\n", 'a' + col, 8 - row);
                led_set(col, row, COLOR_ERROR);
                led_refresh();
                return false;
            }
            // If there's no piece in the board array, there should be no magnet detected
            if (board[row][col] == ' ' && !matrix[row][col]) {
```